### ISSUE TITLE
Remove single instance restriction from dependency solver for v2-build

### ISSUE THREAD

**Issue by grayjay** [haskell/cabal#6459]
The dependency solver always avoids choosing multiple instances for the same package and version within an install plan, even though that restriction isn't necessary with v2-build.

The constraint is enforced in a separate tree transformation:

https://github.com/haskell/cabal/blob/7fd2f604135e32a78301e012eddb9acf3dfe3bf9/cabal-install/Distribution/Solver/Modular/Preference.hs#L450-L456

Removing the single instance restriction has several benefits:

- It would make it easier for cabal to find a solution. See #4192 for an example where the single instance restriction caused more backtracking.
- It would allow cabal to solve for different qualified goals more independently.  Removing the single instance restriction would probably be necessary for implementing the install plan caching discussed in  #6309.
- Skipping the tree transformation would improve performance, even when there are no conflicts.  For example, the profile in https://github.com/haskell/cabal/issues/6309#issuecomment-565794421 shows that `enforceSingleInstanceRestriction` took 4.5% of the time.
- It might improve reuse of packages across install plans by not causing the solver to pick a less preferred version simply because the preferrred version was already used within the install plan.

One disadvantage of removing the single instance restriction is that it would complicate the install plan by making it so that an instance is no longer uniquely identified by its package name and version.  We might need to add more information when logging the install plan or errors.

Removing the single instance restriction requires improving the code that converts the solver's install plan to the install plan used by new-build, such as this TODO:

https://github.com/haskell/cabal/blob/7fd2f604135e32a78301e012eddb9acf3dfe3bf9/cabal-install/Distribution/Solver/Modular/Assignment.hs#L41-L47

**Comment by phadej** [comment]
I'm not sure I follow, it's the case that solver does pick different versions for packages already, e.g. even building `cabal` project, we have local `Cabal-3.3`, and `Cabal-3.0.0.0` somewhere for `build-type: Custom`.

Or is there a restriction that all qualified goals would pick the same `Cabal-3.0.0.0` now?

**Comment by grayjay** [comment]
Yes, the solver just forces Cabal-3.0.0.0 under all qualifiers to be the same instance (have the same version and exactly the same dependencies), so that Cabal-3.0.0.0 is only built once.
